loadtest/control/simulcontroller/websocket.go

```diff
@@ -13,51 +16,79 @@ import (
 // This is used to model user behaviour by responding to certain events with
 // the appropriate actions. It differs from userentity.wsEventHandler which is
 // instead used to manage the internal user state.
-func (c *SimulController) wsEventHandler() {
+func (c *SimulController) wsEventHandler(wg *sync.WaitGroup) {
+	semCount := runtime.GOMAXPROCS(0) * 8
+	semaphore := make(chan struct{}, semCount)
+
+	defer func() {
+		for i := 0; i < semCount; i++ {
+			semaphore <- struct{}{}
+		}
+		wg.Done()
+	}()
+
 	for ev := range c.user.Events() {
 		switch ev.EventType() {
 		case model.WEBSOCKET_EVENT_TYPING:
-			userId, ok := ev.Data["user_id"].(string)
-			if !ok || len(userId) == 0 {
+			userId, ok := ev.GetData()["user_id"].(string)
+			if !ok || userId == "" {
 				c.status <- c.newErrorStatus(fmt.Errorf("simulcontroller: invalid data found in event data"))
 				break
 			}
-
 			user, err := c.user.Store().GetUser(userId)
 			if err != nil {
 				c.status <- c.newErrorStatus(fmt.Errorf("simulcontroller: GetUser failed %w", err))
 				break
 			}
 
 			// The user was found, we check if we have the status for it.
-			if len(user.Id) != 0 {
+			if user.Id != "" {
 				status, err := c.user.Store().Status(userId)
 				if err != nil {
 					c.status <- c.newErrorStatus(fmt.Errorf("simulcontroller: Status failed %w", err))
 					break
 				}
 
 				// If we can't find the user status in the store we fetch it.
-				if len(status.UserId) == 0 {
-					if err := c.user.GetUsersStatusesByIds([]string{user.Id}); err != nil {
-						c.status <- c.newErrorStatus(fmt.Errorf("simulcontroller: GetUsersStatusesByIds failed %w", err))
-						break
+				if status.UserId == "" {
+					select {
```

Comment:
I would introduce a separate buffered channel here, so that we have separate control over the websocket consume rate, and the concurrency of the counting semaphore.

The semaphore concurrency can be `cpu*8` .. that's okay. But this also means that we can only have cpu*8 in-flight requests. And that's not good. Optimizing one will mess up the other. That's why I'd suggest to have a separate buffered channel with a higher buffer - let's say 1/8th of the server's broadcast queue size 4096/8= 512 (can be something else too .. 0/5). And we do the same thing here too, do a non-blocking send with a default clause if the buffer is full.

And in the receiver end, we have the actual semaphore implementation, where it does the token getting/releasing mechanism.

This way we have separated the 2 knobs and have optimal control over both. 

P.S.- This applies to the other other `case semaphore <- struct{}{}:` too obviously :P

Reply:
Discussed offline. `EventChannel` is already buffered. There is no use of using yet another buffered channel.